frontend/src/metabase/lib/api.js

```diff
@@ -36,7 +36,26 @@ export class Api extends EventEmitter {
   apiKey = "";
   sessionToken;
 
-  onBeforeRequest;
+  /**
+   * @typedef {{
+   *    method: "GET" | "POST";
+   *    url: string;
+   *    options: {
+   *      headers?: Record<string, string>;
+   *      hasBody: boolean;
+   *    } & Record<string, unknown>;
+   * }} OnBeforeRequestHandlerData
+   *
+   * @typedef {(data: OnBeforeRequestHandlerData) => (Promise<void | OnBeforeRequestHandlerData>)} OnBeforeRequestHandler
+   *
+   * @typedef {{
+   *   key: string;
+   *   handler: OnBeforeRequestHandler
+   * }} OnBeforeRequestHandlerDescription
+   *
+   * @type {OnBeforeRequestHandlerDescription[]}
+   */
+  onBeforeRequestHandlers = [];
```

Comment:
nit: if each request handler has a key, then I would usually use a map there so its `Map<Key, HandlerFn>`, then we can iterate on `map.values`